cmd/thanos/compact.go

```diff
@@ -462,6 +462,11 @@ func runCompact(
 	}
 
 	compactMainFn := func() error {
+		// this should happen before any compaction to remove unnecessary process on backlogs beyond retention.
+		if err := compact.ApplyRetentionPolicyByTenant(ctx, logger, insBkt, sy.Metas(), retentionByTenant, compactMetrics.blocksMarked.WithLabelValues(metadata.DeletionMarkFilename, metadata.TenantRetentionExpired)); err != nil {
+			return errors.Wrap(err, "retention by tenant failed")
+		}
```

Comment:
Do we have a feature flag to turn it off in case it goes wrong?

Reply:
yes, it will be disabled by default, only when you add `--retention.tenant` which is controlled by shard-flags already